feat: support multiple login modules

tomaspemora opened the pull request:
## Descripción

Permite que el módulo sea utilizado en conjunto con la aplicación [eol_sso_login](https://github.com/eol-uchile/eol_sso_login/tree/master/eol_sso_login) (Login OpenUchile) además de la que ya soportaba [uchiledxlogin](https://github.com/eol-uchile/uchileedxlogin/blob/master/uchileedxlogin/views.py) (Login EOL). 

Los cambios se refieren principalmente en cuanto a la importación del módulo de login así como también a la implementación de una capa de abstracción para la interacción con este.

catalinaIlabaca commented:
Se necesita modificar el test.sh para reflejar la nueva opción 
![imagen](https://github.com/user-attachments/assets/cf598600-93a3-4735-bd1b-39232abfc3d4)

Falta tambien una manera de llegar al las ,lineas 35-41 en las views, pues al no hacerlo, el coverage del test baja de 100 a 99.08

catalinaIlabaca commented on sence/views.py:
```diff
@@ -12,7 +12,10 @@
-from uchileedxlogin.models import EdxLoginUser
+from django.apps import apps
+
+from datetime import datetime
```
Esta importación ya se encuentra en el código en la linea 4

catalinaIlabaca commented on sence/views.py:
```diff
@@ -12,7 +12,10 @@
 from django.urls import reverse
 from django.views.decorators.csrf import csrf_exempt
-from uchileedxlogin.models import EdxLoginUser
+from django.apps import apps
```
Esta importación debe quedar primera en este grupo, pues no esta en orden

catalinaIlabaca commented on sence/views.py:
```diff
@@ -5,14 +5,15 @@
 from django.urls import reverse
 from django.views.decorators.csrf import csrf_exempt
-from uchileedxlogin.models import EdxLoginUser
+
```
eliminar linea vacia

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,65 @@
+    try:
+        login_obj = LoginUserModel.objects.get(user=user)
+        raw = _raw_run(login_obj)
+        return _format_run(raw)
```
porque no cambiarle el nombre y llamar directamente a format_run?

tomaspemora replied:
`_format_run_` es el metodo privado y `format_run` el público. Solo por eso dejé los 2, por "completitud"

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,65 @@
+    return "{}-{}".format(aux[:-1], aux[-1:])  # add '-' before last digit
+
+def _raw_run(obj):
+    if MODEL_USED == 'uchileedxlogin':
```
agregar descripcion

catalinaIlabaca commented on .github/test.sh:
```diff
@@ -2,14 +2,15 @@
 set -e
 
-pip install --src /openedx/venv/src -e git+https://github.com/eol-uchile/uchileedxlogin@0.0.1#egg=uchileedxlogin
+# pip install --src /openedx/venv/src -e git+https://github.com/eol-uchile/uchileedxlogin@0.0.1#egg=uchileedxlogin
```
eliminar codigo comentado

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,65 @@
+import logging
```
agregas categorias de importacion

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,65 @@
+        kwargs['type'] = type_value or DEFAULT_TYPE
+    return kwargs
+
+
```
eliminar lineas vacias

catalinaIlabaca commented on sence/tests.py:
```diff
@@ -8,7 +8,8 @@
 from django.urls import reverse
 from mock import patch, Mock
-from uchileedxlogin.models import EdxLoginUser
+logger = logging.getLogger(__name__)
```
mover le logger hasta después de las importaciones

catalinaIlabaca commented on sence/views.py:
```diff
@@ -3,6 +3,7 @@
 from __future__ import unicode_literals
 from datetime import datetime
 import logging
+logger = logging.getLogger(__name__)
```
porque cambiaste el logger de lugar?  en la primera parte del codigo la idea es que sol hayan importaciones

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,65 @@
+        logger.warning("{} doesn't have RUN".format(user.username))
+        return ''
+
+def format_run(run):
```
agregar descripcion

fsologureng commented on sence/login_interface.py:
```diff
@@ -0,0 +1,65 @@
+HAS_TYPE    = any(f.name == 'type' for f in LoginUserModel._meta.get_fields())
+DEFAULT_TYPE = 'rut'
+
+def _format_run(run):
```
No puedo dejar de asociar la palabra run con su significado en inglés la primera vez que lo leo, lo cual creo que llama a confusión. Creo que sería mejor usar una palabra como *document_id* o *doc_id*, no sé, algo más genérico, o por último *rut* para evitar la confusión con el inglés.

fsologureng commented on sence/tests.py:
```diff
@@ -105,19 +103,19 @@ def test_format_run(self):
+        new_run_4 = format_run(run_4)
         self.assertEqual(new_run_4, '1234567-K')
 
     def test_get_user_run(self):
```
Esta función todavía tiene una mención a uchileedxlogin en su descripción.

fsologureng commented on sence/tests.py:
```diff
@@ -105,19 +103,19 @@ def test_format_run(self):
             Run from uchileedxlogin are in the format: 00123456789
```
mención a uchileedxlogin

fsologureng reviewed (changes requested):
En primer lugar, estoy de acuerdo con Cata en resolver los temas de formato porque ella ya hizo la pega en ese sentido y sería bueno mantenerla. También con agregar descripciones a los métodos de la nueva interfaz abstracta para los logins.

En segundo lugar, lo pedido en cuanto a cambiar el uso de *run* como parte de los nombre en inglés, si te parece. Probablemente quede hacer lo mismo con el propio paquete y con la separación entre vista y controlador (todos los métodos get_ y set_ deberían estar fuera de view), pero eso otro sería en otro pr.

Por último, todavía hay algunas referencias al paquete `uchileedxlogin` fuera de la nueva interfaz, que sería bueno eliminar.

Finalmente, dejaría 2 commits para el pr via rebase/squash: el refactor del código para agregar la nueva interfaz todo en 1 commit, y el commit que adapta el test.sh para probar con los 2 paquetes de login.

fsologureng commented:
Comentario aparte: Se extraña que no hay tests de JS, que es donde finalmente se hace toda la pega de probar la conexión a SENCE.

catalinaIlabaca commented on sence/tests.py:
```diff
@@ -498,7 +509,7 @@ def test_login_sence(self):
 
         # 4
-        EdxLoginUser.objects.create(user=self.user, run='000012345678')
+        # create_sso_user(self.user, document_value='000012345678')
```
codigo comentado eliminar

tomaspemora replied:
debes esperar a que te pida review de nuevo. Este cambio lo abordé en mi último commit.

catalinaIlabaca commented on sence/tests.py:
```diff
@@ -99,40 +97,51 @@ def setUp(self):
+        self.assertEqual(user_document_id_2, '123456-7')
+        self.assertEqual(user_document_type_2, 'rut')
+
+        if MODEL_USED == 'eol_sso_login':
```
@tomaspemora el MODEL_USED que seteas en login_inferface.py es eolssologin, esta linea nunca es alcanzada por ninguno de los dos test, debes  o modificar esta linea o el  MODEL_USED='eol_sso_login' en  login_inferface.py

tomaspemora replied:
Fíjate bien, hay 2 casos para `MODEL_USED`: `uchileedxlogin` o `eol_sso_login` dependiendo de cuál módulo tengo instalado.

tomaspemora replied:
corregido

tomaspemora commented:
Agregué 4 commits para hacerme cargo de los cambios a los que aludieron. Ordené las importaciones y respeté las categorías definidas por Cata. Cambié el concepto `run` por `document_id` en todas las partes del código donde lo encontré. Ahora `run` solo representa el caso especial para el `document_id` de `uchileedxlogin`. 

Generé 2 badges de coverage en este caso, una para cada módulo de login. Por esto, tuve que cambiar un poco `test.sh`. Aproveché de subir la versión en el `setup.py`.

Una vez que esté aprobado, voy a hacer el squash para dejar los 2 commits mencionados por Felipe.

fsologureng commented on .github/test.sh:
```diff
@@ -2,17 +2,30 @@
-ln -s /openedx/staticfiles ./test_root/
-
-DJANGO_SETTINGS_MODULE=lms.envs.test EDXAPP_TEST_MONGO_HOST=mongodb pytest sence/tests.py
+rm -f coverage-badge-*.svg .coverage*
```
no corresponde borrar un recurso que está versionado

fsologureng commented on sence/tests.py:
```diff
@@ -99,40 +97,51 @@ def setUp(self):
+        self.assertEqual(user_document_id_2, '123456-7')
+        self.assertEqual(user_document_type_2, 'rut')
+
+        if MODEL_USED == 'eol_sso_login':
```
Creo que es mala idea hacer explícito el modelo acá porque rompes con la idea del encapsulamiento de la interfaz que desarrollaste. Eso empieza a revelarse en el coverage, lo cual no corresponde porque al comportamiento del módulo para sence propiamente tal no debería importarle qué biblioteca de login subyacente se usa. Para eso la interfaz abstracta debería estar fuera del coverage, porque es un parche, no para sence, si no para la ausencia de API en las implementaciones del login; Si aquellas fueran implementaciones interoperables, ellas mismas deberían implementar los métodos comunes para que paquetes como estos los usen, y sence se comportaría sin distinción con ellas.
Además, entendí que SENCE no soporta otro tipo de document id, por tanto no tiene sentido testear otros tipos __en este paquete__.

tomaspemora replied:
perfecto! Sí, también pensé que esta parte del test podría estar demás. En esa misma línea, creo que eliminaré otros tests como [`test_make_login_kwargs_with_type` ](https://github.com/eol-uchile/sence-xblock/blob/tpeet/flexible-sso-module/sence/tests.py#L654) o [`test_format_document_id_non_rut`](https://github.com/eol-uchile/sence-xblock/blob/tpeet/flexible-sso-module/sence/tests.py#L648)

tomaspemora replied:
Voy a omitir del coverage los test a `login_interface.py`

fsologureng commented on sence/login_interface.py:
```diff
@@ -0,0 +1,81 @@
+    return kwargs
+
+# public methods
+def create_sso_user(user, document_value=None, type_document='rut'):
```
usar `document_id` en vez de _value por fa, se entiende muy raro, aunque la llave se llame igual que el objeto. Tb puede ser `doc_id` para diferenciar del objeto "Document ID"

fsologureng replied:
user `document_type` (o `doc_type`) para alinear sintaxis con el otro parámetro (no spanglish).

fsologureng commented on sence/tests.py:
```diff
@@ -99,40 +97,51 @@ def setUp(self):
         """
-            Test format run to sence requirements (123456-7)
-            Run from uchileedxlogin are in the format: 00123456789
+            Test format document_id to sence requirements (123456-7)
```
Este test está planteado totalmente desde la lógica uchileedxlogin; El formato 00123456789 no es un formato que provea el módulo sence, es un formato que provee el paquete uchileedxlogin, por tanto este test estaba verificando que el paquete uchileedxlogin puede generar un document_id en la lógica que sence necesita. Es un test que no es parte del negocio de este paquete, a lo más de la interfaz abstracta.
Lo que este paquete (módulo) necesita es verificar que, dado un usuario, al recuperarse su rut (doc id), este debiera entregarlo en el formato correcto para sence, sea cual sea el login subyacente. Ese es el test.

Dicho lo anterior, el método `format_document_id` no se necesita como método público, porque no es negocio de sence entregar el RUT en el formato que lo requiere si no solo de la interfaz abstracta parche.

fsologureng reviewed (changes requested):
Puchas lo siento porque veo que hiciste N esfuerzo por diferenciar los 2 comportamientos con los 2 logins subyacentes. Sin embargo, creo que es mala idea porque refuerza la dependencia granular de este paquete con los otros 2; Cada cambio en esas dependencias, fácilmente implicará cambios acá, por eso la importancia de definir una interfaz común lo más rígida posible que permita integrarse con ambos sin distinción de cual se usa, porque el negocio de este paquete no es arreglar a los otros si no solo pedirles el RUT del usuario. No hay más necesidad que esa, y en ese sentido, se está creando la interfaz abstracta para desacoplar sence con uchileedxlogin, pero al mismo tiempo para que el módulo pueda ser usado con "cualquiera" que provea ese RUT.

Ahora, entiendo que habrá casos de uso que no se podrían verificar bajo esa lógica, como el de un usuario con DNI en vez de RUT, pero en la práctica no es distinto del comportamiento de un usuario que no tiene documento de identidad, lo cual puede pasar con ambos logins subyacentes; Ambos deben considerar qué hacer si se les pide el rut de esos usuarios. La idea es encapsular **todas** las diferencias en la interfaz abstracta creando una respuesta uniforme, sin distinción.

catalinaIlabaca commented on sence/tests_login_interface.py:
```diff
@@ -0,0 +1,123 @@
+        result = create_sso_user(self.user, document_id='00001234567')
+        self.assertFalse(result)
+
+    def test_create_sso_user_with_dni(self):
```
Agregar  descripción

catalinaIlabaca commented on sence/tests.py:
```diff
@@ -22,11 +21,11 @@
 from . import views
 from .models import EolSenceStudentStatus, EolSenceStudentSetup, EolSenceCourseSetup
 from .sence import SenceXBlock, get_configurations, get_students_setups
+from .login_interface import create_sso_user
```
Mover al orden alfabetico (deberia quedar debajo de from . import views)
o aplicar lo siguiente
En VS seleccionar las lineas que quieres dejar en orden alfabético y  ctrl+shift+p y luego buscar sort lines ascending

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+# Python Standard Libraries
+import logging
+logger = logging.getLogger(__name__)
```
Esta definición debe ir al final de las importaciones,  de decir, des pues de la linea 7

catalinaIlabaca commented on sence/login_interface.py:
```diff
@@ -0,0 +1,80 @@
+
+# Installed packages (via pip)
+from django.apps import apps
+if apps.is_installed('uchileedxlogin'):
```
dejar un espacio para orden del código

catalinaIlabaca commented on sence/views.py:
```diff
@@ -22,6 +21,7 @@
 
 # Internal project dependencies
 from .models import EolSenceCourseSetup, EolSenceStudentSetup, EolSenceStudentStatus
+from .login_interface import get_user_document_id_and_type
```
Por orden alfabético, login_interface debe ir primero

catalinaIlabaca reviewed (changes requested):
Son cambios pequeños de formato y agregar una descripción

tomaspemora commented:
entrego la posta del PR a @jjorgeb, ya que el estará viendo la integración de todos los módulos con `uchileedxlogin`

tomaspemora commented on sence/integrations/login_interface.py:
```diff
@@ -0,0 +1,91 @@
+        return document_id
+
+
+def _raw_document_id_and_type(obj):
```
esta función ya no se utiliza para `uchileedxlogin` y solo queda utilizable el segundo caso. Deberíamos eliminar la función, para ir reduciendo esta interfaz.

tomaspemora commented:
También considerar instalar el modulo uchileedxlogin en un release dev para que corran los tests.

catalinaIlabaca commented on README.MD:
```diff
@@ -1,8 +1,6 @@
-![Coverage Status](/coverage-badge.svg)
-
-![https://github.com/eol-uchile/sence-xblock/actions](https://github.com/eol-uchile/sence-xblock/workflows/Python%20application/badge.svg)
+![https://github.com/eol-uchile/eol_survey/actions](https://github.com/eol-uchile/sence-xblock/workflows/Python%20application/badge.svg)
```
DUDA: porque estamos haciendo referencia a survey

jjorgeb replied:
buena pregunta. me imagino que fue un error

jjorgeb commented:
@fsologureng @tomaspemora listo para revision

fsologureng reviewed (approved):
No revisé la lógica del paquete, si no que me fijé en el refactor y su estructura.

Encuentro que la interfaz propuesta es muy reusable para un paquete dedicado que haga de wrapper de los 2 paquetes para el sso, y la aprovecharía para ello.

Queda por hacer el squash de los commits por tipo.

catalinaIlabaca commented on README.MD:
```diff
@@ -1,7 +1,5 @@
 # SENCE XBLOCK
 
-![Coverage Status](/coverage-badge.svg)
```
Hay que mantener el coverage

tomaspemora commented on sence/integrations/login_interface.py:
```diff
@@ -0,0 +1,56 @@
+    from uchileedxlogin.services.interface import get_doc_id_by_user_id
+    from uchileedxlogin.services.utils import get_document_type
+    MODEL_USED = 'uchileedxlogin'
+    ADMITTED_DOCUMENT_TYPES = ['rut']
```
esto no se está usando?

jjorgeb replied:
no

tomaspemora commented on sence/integrations/login_interface.py:
```diff
@@ -0,0 +1,56 @@
+    from uchileedxlogin.services.utils import get_document_type
+    MODEL_USED = 'uchileedxlogin'
+    ADMITTED_DOCUMENT_TYPES = ['rut']
+    DOCUMENT_FIELD = 'run'
```
esto tampoco

jjorgeb commented:
@tomaspemora

tomaspemora commented:
Aprobado

fsologureng commented:
@jjorgeb squash adecuado please

jjorgeb committed:
feat: adds support for multiple login modules and updates compatibility with new version of uchileedxlogin

- adds login_interface.py and api.py for handling login with different user models.
- updates compatibility with new version of uchileedxlogin, making use of its service layer.

jjorgeb committed:
test: updates tests and adds tests for both new feafures and to test login_interface

- Modifies test that needed some users to be associate with a doc_id to mock that behaviour instead of creating an user.
- Splits some test functions for easier testing.
- Adds new test functions.
- Adds test_login_interface.py to tests the new login interface.

jjorgeb committed:
build: bumps version to 1.1.0

jjorgeb committed:
ci: updates test script to test both login interfaces and updates uchileedxlogin version used for testing

- removes integrations folder from coverage

jjorgeb commented:
listo el squash

fsologureng commented on sence/integrations/tests_login_interface.py:
No entiendo por qué es necesario identificar el modelo si se supone que el objetivo de la interfaz es enmascarar los modelos detrás y ya lo está haciendo dentro de cada uno de sus métodos.

Por otro lado, si fuese necesario testear la propia interfaz (que parece ser el objetivo de este set de tests), entonces sería mejor separar el set para cada modelo en archivos diferentes y reflejar eso en la llamada a pytest con cada uno cuando corresponda (el commit del ci).
El primer test, que parece ser genérico, debiera moverse al tests.py genérico. Los demás parecen ser los tests sólo para uchileedxlogin, por tanto es más consistente un set de tests específico para ese caso, aunque conservo la duda de si cada método ya sabe qué está instalado y no influye preguntar qué modelo está instalado para que funcionen bien.

Entonces, o la interfaz enmascara a los modelos y en verdad los tests no debieran identificar el modelo, o los tests son específicos para probar la propia interfaz (y no el paquete sence) y en ese es más adecuado tener sets diferentes porque los tests serán diferentes según cada instalación.

fsologureng reviewed (changes requested):
Ahora que veo el squash veo con más claridad la propuesta de cambios. Creo que hay que darle una vuelta más a la estructura de los tests de la interfaz.

jjorgeb committed:
test: splits interface testing and fixes eol_sso_login testing

jjorgeb committed:
test: delete unused test file

jjorgeb commented:
Se me habia pasado eliminar el archivo de test que separe en 2

jjorgeb commented:
Estoy trabajando en la implementacion directa con eol_sso. No revisar.

tomaspemora commented:
La funciónalidad se implementó en https://github.com/eol-uchile/sence-xblock/pull/11
